src/channel.cpp

```diff
@@ -518,11 +518,85 @@ namespace mamba
+     ************************************/
+
     Channel& make_channel(const std::string& value)
     {
         return Channel::make_cached_channel(value);
     }
 
+    void append_channel_urls(const std::string name,
+                             const std::vector<std::string>& platforms,
+                             bool with_credential,
+                             std::vector<std::string>& result,
+                             std::set<std::string>& control)
+    {
+        bool ret = !control.insert(name).second;
+        if (ret)
+            return;
+
+        std::vector<std::string> urls = make_channel(name).urls(platforms, with_credential);
+        std::copy(urls.begin(), urls.end(), std::back_inserter(result));
+    }
+
+    std::vector<std::string> get_channel_urls(const std::vector<std::string>& channel_names,
+                                              const std::vector<std::string>& platforms,
+                                              bool with_credential)
+    {
+        std::set<std::string> control;
+        std::vector<std::string> result;
+        result.reserve(channel_names.size() * platforms.size());
+        for (const auto& name: channel_names)
+        {
+            auto multi_iter = ChannelContext::instance().get_custom_multichannels().find(name);
+            if (multi_iter != ChannelContext::instance().get_custom_multichannels().end())
+            {
+                for (const auto& n: multi_iter->second)
+                {
+                    append_channel_urls(n, platforms, with_credential, result, control);
+                }
+            }
+            else
+            {
+                append_channel_urls(name, platforms, with_credential, result, control);
+            }
+        }
+        return result;
+    }
+
+    std::vector<std::string> calculate_channel_urls(const std::vector<std::string>& channel_names,
+                                                    bool prepend,
+                                                    const std::string& platform,
+                                                    bool use_local)
+    {
+        std::vector<std::string> platforms = platform.size()
+                                           ? std::vector<std::string>({ platform, "noarch" })
+                                           : Context::instance().platforms();
+
+        if (prepend || use_local)
+        {
+            const auto& ctx_channels = Context::instance().channels;
+            std::vector<std::string> names;
+            names.reserve(channel_names.size() + 1 + ctx_channels.size());
+            if (use_local)
+            {
+                names.push_back(LOCAL_CHANNELS_NAME);
+            }
+            std::copy(channel_names.begin(), channel_names.end(), std::back_inserter(names));
+            if (prepend)
+            {
+                std::copy(ctx_channels.begin(), ctx_channels.end(), std::back_inserter(names));
+            }
```

Comment:
what happens if prepend is false?

Reply:
ah so the opposite of prepend is not append, it's just to ignore the ctx_channels?

Reply:
Nothing.

Reply:
Btw I kept the name from conda, but I think we should rename it into something more explicit, like "append_context_channel".

Also in our case, the channels_names and the ctx_channels are exactly the same (the context always contains those from the command line + configuration), and I wonder if we'll have a use case where we pass different channel names than those from the command line.